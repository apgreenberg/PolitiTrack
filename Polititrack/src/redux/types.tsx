import { UserInfo } from '../model/database/userInterfaces';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { store } from './store';

//Action types
export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';

export interface LoginAction {
  type: typeof LOGIN;
  payload: UserInfo;
}

export interface LogoutAction {
  type: typeof LOGOUT;
}

export type Actions = LoginAction | LogoutAction;

//State interface typing
export interface StateWrapper {
  loggedIn: boolean;
  currentUser: UserInfo;
}

//Dispatch types
export type ThunkActionType<ReturnType = void> = ThunkAction<
  ReturnType,
  StateWrapper,
  unknown,
  Actions
>;

export type ThunkDispatchType = ThunkDispatch<StateWrapper, undefined, Actions>;

export type AppDispatch = typeof store.dispatch;
