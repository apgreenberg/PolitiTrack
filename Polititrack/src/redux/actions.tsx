import { ThunkActionType } from './types';
import { ActionCreator, Dispatch } from 'redux';
import {
  attemptSignIn,
  attemptSignOut,
  attemptDeleteAccount,
  findUser,
  pullCurrentUserInfo,
} from '../model/database/handleLogin';
import { getNextId, addUser } from '../model/database/handleSignUp';
import { updateInternalDatabase } from '../model/database/userHelpers';
import { UserInfo } from '../model/database/userInterfaces';
import {
  initializeNotifications,
  removeNotifications,
} from '../model/background/notifications';

/**
 * Checks if the user exists and if so signs them in
 * @param email - the user's email
 * @param password - the user's password
 * @returns - An action to modify the logged in state
 */
export const signIn: ActionCreator<ThunkActionType> = (
  email: string,
  password: string
) => {
  return async (dispatch: Dispatch) => {
    const foundUserInfo: UserInfo | null = findUser(email, password);
    if (foundUserInfo !== null) {
      await attemptSignIn(foundUserInfo);
      await initializeNotifications(foundUserInfo);
      dispatch({ type: 'LOGIN', payload: foundUserInfo });
    } else {
      alert('User does not exist!');
    }
  };
};

/**
 * Signs out the current user
 * @returns - an action to modify the logged in state
 */
export const signOut: ActionCreator<ThunkActionType> = () => {
  return async (dispatch: Dispatch) => {
    await attemptSignOut();
    await removeNotifications();
    dispatch({ type: 'LOGOUT' });
  };
};

/**
 * Signs up a new user and logs them in
 * @param email - the new user's email
 * @param password - the new user's password
 * @returns - an action to modify the logged in state
 */
export const signUp: ActionCreator<ThunkActionType> = (
  email: string,
  password: string
) => {
  return async (dispatch: Dispatch) => {
    const id: number = getNextId();
    const newUser: UserInfo = {
      id: id,
      token: 'abc',
      email: email,
      password: password,
      name: 'default name',
      data: { jurisdictions: [], bills: [], billTypes: [], legislators: [] },
      settings: { updateJurisdictions: false, updateBills: false },
    };
    await addUser(newUser);
    await attemptSignIn(newUser);
    await initializeNotifications(newUser);
    dispatch({ type: 'LOGIN', payload: newUser });
  };
};

/**
 * Retrieves the past user from AsyncStorage if there is one
 * @returns - an action to modify the logged in state
 */
export const retrievePastUser: ActionCreator<ThunkActionType> = () => {
  return async (dispatch: Dispatch) => {
    await updateInternalDatabase();
    const userInfo: UserInfo | null = await pullCurrentUserInfo();
    if (userInfo !== null) {
      await initializeNotifications(userInfo);
      dispatch({ type: 'LOGIN', payload: userInfo });
    }
  };
};

/**
 * Logs out and deletes the user with the associated id
 * @param id - the user to delete
 * @returns - an action to modify the logged in state
 */
export const deleteAccount: ActionCreator<ThunkActionType> = (id: number) => {
  return async (dispatch: Dispatch) => {
    await attemptSignOut();
    await removeNotifications();
    await attemptDeleteAccount(id);
    dispatch({ type: 'LOGOUT' });
  };
};
