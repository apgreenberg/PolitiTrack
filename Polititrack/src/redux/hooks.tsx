import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { UserInfo } from '../model/database/userInterfaces';
import type { StateWrapper, AppDispatch } from './types';

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<StateWrapper> = useSelector;
export const selectLoginState = (state: StateWrapper): boolean =>
  state.loggedIn;
export const selectCurrentUser = (state: StateWrapper): UserInfo =>
  state.currentUser;
