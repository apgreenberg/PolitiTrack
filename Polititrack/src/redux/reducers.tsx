import { blankUserInfo } from '../model/database/userInterfaces';
import { Actions, LOGIN, LOGOUT, StateWrapper } from './types';

const initialState: StateWrapper = {
  loggedIn: false,
  currentUser: blankUserInfo,
};

const reducer = (
  state: StateWrapper = initialState,
  action: Actions
): StateWrapper => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        loggedIn: true,
        currentUser: action.payload,
      };
    case LOGOUT:
      return {
        ...state,
        loggedIn: false,
      };
    default:
      return state;
  }
};

export default reducer;
