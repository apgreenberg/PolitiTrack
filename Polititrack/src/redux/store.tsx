import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import reducer from './reducers';
import { ThunkDispatchType, StateWrapper } from './types';

export const store = createStore(
  reducer,
  applyMiddleware<ThunkDispatchType, StateWrapper>(thunk)
);
