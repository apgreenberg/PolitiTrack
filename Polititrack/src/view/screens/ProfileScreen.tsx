import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  SafeAreaView,
  Button,
  Alert,
} from 'react-native';
import 'react-native-gesture-handler';
import { useIsFocused } from '@react-navigation/native';

import CustomButton from '../components/CustomButton';
import text from '../styles/text';

import { signOut, deleteAccount } from '../../redux/actions';
import {
  selectCurrentUser,
  useAppSelector,
  useAppDispatch,
} from '../../redux/hooks';

import { UserInfo } from '../../model/database/userInterfaces';
import { Dispatch, Action } from 'redux';
import { ThunkDispatchType } from '../../redux/types';

/**
 * Displays the profile screen, inluding the signOut button
 * @returns - the functional component representing the Profile screen
 */
const ProfileScreen: React.FC = () => {
  const dispatch: Dispatch<Action> & ThunkDispatchType = useAppDispatch();
  const currentUser: UserInfo = useAppSelector(selectCurrentUser);
  const focused = useIsFocused();
  React.useEffect(() => {}, [focused]);

  const deleteAccountAlert = () =>
    Alert.alert('Are you sure?', 'This cannot be undone', [
      {
        text: 'Cancel',
        onPress: () => {},
        style: 'cancel',
      },
      { text: 'OK', onPress: () => dispatch(deleteAccount(currentUser.id)) },
    ]);
  return (
    <SafeAreaView>
      <View style={{ flexDirection: 'column' }}>
        <View style={{ height: '20%', padding: 5 }}>
          <Text style={text.largeText}>Hi, {currentUser.name}!</Text>
          <Text style={text.mediumText}>Email: {currentUser.email}</Text>
        </View>
        <View style={styles.buttonPanelAndLogoutContainer}>
          <View style={styles.buttonPanel}>
            {['My Bills', 'My Legislators', 'My Jurisdictions'].map(
              (item: string, key: number) => (
                <View style={styles.buttonContainer} key={key}>
                  <CustomButton type={item}></CustomButton>
                </View>
              )
            )}
          </View>
          <View style={styles.signOutButtonContainer}>
            <Button
              testID={'signOutButton'}
              title="Sign Out"
              onPress={() => dispatch(signOut())}
            ></Button>
            <Button
              testID={'deleteAccountButton'}
              title="Delete Account"
              onPress={() => deleteAccountAlert()}
            ></Button>
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  buttonPanelAndLogoutContainer: {
    height: '80%',
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  buttonPanel: {
    flexShrink: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  buttonContainer: {
    alignItems: 'center',
    height: '33%',
    width: '100%',
  },
  signOutButtonContainer: {
    paddingTop: 60,
    paddingBottom: 20,
    justifyContent: 'flex-end',
  },
});
export default ProfileScreen;
