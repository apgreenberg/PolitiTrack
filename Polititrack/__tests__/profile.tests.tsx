import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import ProfileScreen from '../src/view/screens/ProfileScreen';
import { render, fireEvent } from '@testing-library/react-native';

import { useSelectorMock, userData } from './common';

import * as actions from '../src/redux/actions';
const signOutMock = jest.spyOn(actions, 'signOut');

/**
 * Tests the Landing screen operates correctly if the user is signed in/signed out
 */
describe('Profile tests', () => {
  test('Viewing the profile page', () => {
    useSelectorMock.mockReturnValue(userData);
    const { toJSON } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <ProfileScreen />
      </MockedProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
  test('Test pressing the sign out button', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <ProfileScreen />
      </MockedProvider>
    );
    fireEvent(getByTestId('signOutButton'), 'onPress');
    expect(signOutMock).toHaveBeenCalled();
  });
});
