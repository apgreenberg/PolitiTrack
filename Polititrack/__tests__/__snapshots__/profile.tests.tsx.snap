// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile tests Viewing the profile page 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "height": "20%",
          "padding": 5,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 24,
            "textAlign": "center",
          }
        }
      >
        Hi, 
        default name
        !
      </Text>
      <Text
        style={
          Object {
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        Email: 
        ari.greenberg99@gmail.com
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "height": "80%",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexShrink": 1,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": "33%",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#dae8eb",
                "borderBottomWidth": 2,
                "borderColor": "#9ca7b8",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 2,
                "height": "90%",
                "justifyContent": "center",
                "opacity": 1,
                "width": "90%",
              }
            }
            testID="customButton"
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "textAlign": "center",
                }
              }
            >
              My Bills
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "height": "33%",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#dae8eb",
                "borderBottomWidth": 2,
                "borderColor": "#9ca7b8",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 2,
                "height": "90%",
                "justifyContent": "center",
                "opacity": 1,
                "width": "90%",
              }
            }
            testID="customButton"
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "textAlign": "center",
                }
              }
            >
              My Legislators
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "height": "33%",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#dae8eb",
                "borderBottomWidth": 2,
                "borderColor": "#9ca7b8",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 2,
                "height": "90%",
                "justifyContent": "center",
                "opacity": 1,
                "width": "90%",
              }
            }
            testID="customButton"
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "textAlign": "center",
                }
              }
            >
              My Jurisdictions
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "justifyContent": "flex-end",
            "paddingBottom": 20,
            "paddingTop": 60,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="signOutButton"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Sign Out
            </Text>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="deleteAccountButton"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Delete Account
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
